src/protocols/secure_channel/CASESession.cpp

```diff
@@ -404,187 +445,283 @@ CHIP_ERROR CASESession::HandleSigmaR1(System::PacketBufferHandle && msg)
-    SuccessOrExit(err = tlvReader.GetBytes(destinationIdentifier, sizeof(destinationIdentifier)));
+    if (sessionResumptionRequested && memcmp(resumptionID, mResumptionId, kCASEResumptionIDSize) == 0)
+    {
+        // Cross check resume1MIC with the shared secret
+        SuccessOrExit(err = ValidateSigmaResumeMIC(resume1MICSpan, ByteSpan(initiatorRandom), ByteSpan(mResumptionId),
+                                                   ByteSpan(kKDFS1RKeyInfo), ByteSpan(kResume1MIC_Nonce)));
 
+        // Send Sigma2Resume message to the initiator
+        SuccessOrExit(err = SendSigma2Resume(ByteSpan(initiatorRandom)));
+    }
+    else
     {
         const ByteSpan * ipkListSpan = GetIPKList();
         FabricIndex fabricIndex      = Transport::kUndefinedFabricIndex;
         memcpy(mIPK, ipkListSpan->data(), sizeof(mIPK));
+
+        SuccessOrExit(err = tlvReader.Next());
+        VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+        SuccessOrExit(err = tlvReader.GetBytes(destinationIdentifier, sizeof(destinationIdentifier)));
+
         VerifyOrExit(mFabricsTable != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
         fabricIndex = mFabricsTable->FindDestinationIDCandidate(ByteSpan(destinationIdentifier), ByteSpan(initiatorRandom),
                                                                 ipkListSpan, GetIPKListEntries());
         VerifyOrExit(fabricIndex != Transport::kUndefinedFabricIndex, err = CHIP_ERROR_CERT_NOT_TRUSTED);
 
         mFabricInfo = mFabricsTable->FindFabricWithIndex(fabricIndex);
         VerifyOrExit(mFabricInfo != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
-    }
 
-    SuccessOrExit(err = tlvReader.Next());
-    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
-    SuccessOrExit(err = tlvReader.GetBytes(mRemotePubKey, static_cast<uint32_t>(mRemotePubKey.Length())));
+        SuccessOrExit(err = tlvReader.Next());
+        VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+        SuccessOrExit(err = tlvReader.GetBytes(mRemotePubKey, static_cast<uint32_t>(mRemotePubKey.Length())));
+
+        SuccessOrExit(err = SendSigma2());
+    }
 
 exit:
 
-    if (err != CHIP_NO_ERROR)
+    if (err == CHIP_ERROR_CERT_NOT_TRUSTED)
+    {
+        SendStatusReport(mExchangeCtxt, kProtocolCodeNoSharedRoot);
+        mState = kInitialized;
+    }
+    else if (err != CHIP_NO_ERROR)
     {
-        SendErrorMsg(SigmaErrorType::kUnexpected);
+        SendStatusReport(mExchangeCtxt, kProtocolCodeInvalidParam);
+        mState = kInitialized;
     }
     return err;
 }
 
-CHIP_ERROR CASESession::SendSigmaR2()
+CHIP_ERROR CASESession::SendSigma2Resume(const ByteSpan & initiatorRandom)
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
+    size_t data_len = EstimateTLVStructOverhead(kCASEResumptionIDSize + kTAGSize + sizeof(uint16_t), 7);
```

Comment:
Suggest `max_sigma2_resume_data_len`. BTW, the 7 here should be 4, and have `/* + kMRPOptionalParamsLength */`